### ISSUE TITLE
Show warning if .scala or .sbt configuration file changed

### ISSUE THREAD

**Issue by jlcheng** [sbt/sbt#241]
As I am learning SBT, I find myself making changes to .scala configurations and reloading them to test my configurations. Sometimes I forget to type 'reload' in the SBT console and end up being confused at why my last change didn't seem to take effect.

It would be very nice if SBT can "watch" the .sbt and .scala configuration files for change every time a task or command is executed on the SBT console and warn the user of changes, e.g.,

```
> run  
[warn] project/Foo.scala has changed since the last "reload".

Multiple main classes detected, select one to run:

 [1] com.twitter.finagle.example.echo.EchoClient
 [2] com.twitter.finagle.example.echo.EchoServer
```

Or in case more than 1 file was changed:

```
> run
[warn] Configuration files have changed since the last "reload".
```

**Comment by ijuma** [comment]
Yes, this would be very useful.

**Comment by mslinn** [comment]
Is this not similar to 'sbt watch'?

**Comment by jlcheng** [comment]
Perhaps, what is "sbt watch"? The ~reload syntax doesn't detect changes to the SBT configuration files.

**Comment by mslinn** [comment]
Not sure what watch does. This is all I know

```
$ sbt
> tasks
...
  watch                                  Continuous execution configuration.
  watch-sources                          Defines the sources in this project for continuous execution to watch for changes.
  watch-transitive-sources               Defines the sources in all projects for continuous execution to watch.

> watch
[info] sbt.Defaults$$anonfun$watchSetting$1$$anon$1@34c0328
```

**Comment by jlcheng** [comment]
In this case, 'watch' is a setting. See `inspect watch`

```
> inspect watch
[info] Setting: sbt.Watched = sbt.Defaults$$anonfun$watchSetting$1$$anon$1@2aec1c4d
[info] Description:
[info]  Continuous execution configuration.
[info] Provided by:
[info]  {file:/Users/jcheng/work/demo/hello-finagle/}finagle/*:watch
[info] Dependencies:
[info]  *:poll-interval
[info]  *:this-project-ref
[info] Delegates:
[info]  *:watch
[info]  {.}/*:watch
[info]  */*:watch
> inspect poll-interval
[info] Setting: Int = 500
[info] Description:
[info]  Interval between checks for modified sources by the continuous execution command.
[info] Provided by:
[info]  */*:poll-interval
[info] Reverse dependencies:
[info]  *:watch
[info] Delegates:
[info]  *:poll-interval
[info]  {.}/*:poll-interval
[info]  */*:poll-interval
[info] Related:
[info]  */*:poll-interval
```

When you type in `~compile` at the SBT command line, SBT will look for changes every 500 ms (poll-interval) and recompile if something changes. Using the inspect command on watch tells us that you can configure the `watch` setting by setting the poll interval to a different value, e.g.,

```
  pollInterval := 2000
```

Now SBT will look for changes every 2 seconds instead.

So in short, the `watch` setting won't help here.

**Comment by xuwei-k** [comment]
https://gist.github.com/xuwei-k/6278769

**Comment by jsuereth** [comment]
@xuwei-k that's a nice workaround. I think I'd like to tie this to the "new file watching" API we plan to take from play where the notification of change happens asynchronously.

**Comment by xuwei-k** [comment]
https://github.com/tototoshi/sbt-build-files-watcher

**Comment by MasseGuillaume** [comment]
@jsuereth @xuwei-k FYI:
https://github.com/tototoshi/sbt-build-files-watcher/issues/1#issuecomment-72304860

**Comment by dwijnand** [comment]
Fixed in https://github.com/sbt/sbt/pull/4335/commits/28fd4a1e618f938f9e1a4bd9639255c90a629cba#diff-6373e7f7122325e753b75fe1cc76ff5fR391?

**Comment by OlegYch** [comment]
@dwijnand looks like it, will need to try

**Comment by eed3si9n** [comment]
@eatkins How does `watchProjectSources` work? Is it still hooked up?

**Comment by eatkins** [comment]
In my outstanding PRs, I added the feature where changes to *.sbt and *.scala files can automatically trigger a rebuild in a continuous build. We could also add a task that checks for diffs and to the build files and either automatically rebuilds or prints a warning if there are any changes. When to evaluate that task is not entirely obvious. My first thought would be to put it in the compile task, but you might want to automatically reload when running any arbitrary task. That would require running this check before any task evaluation which does add a small overhead. I think that's probably fine as long as it's configurable.

**Comment by eatkins** [comment]
@eed3si9n and I discussed this in today's meeting. The continuous builds in 1.3.0-M3 do watch the project source files and will automatically reload if they've changed. I don't think that it makes sense to have that feature by default if we do not support automatic reloading (or warnings) in regular task evaluation. Given that, I think it makes sense to either put the automatic reloading in continuous builds feature behind a sys prop or to get it working in regular task evaluation before 1.3.0.
